File: mobile-pwa/src/lib/quran.ts
import type { QuranPage } from '@/types'

const BASE_URL = 'https://api.alquran.cloud/v1'
const EDITION = 'ar.uthmani'

// Simple in-memory cache to avoid repeated requests
const pageCache = new Map<number, QuranPage>()

export async function fetchQuranPage(page: number): Promise<QuranPage> {
  if (pageCache.has(page)) return pageCache.get(page)!

  const res = await fetch(`${BASE_URL}/page/${page}/${EDITION}`)
  if (!res.ok) throw new Error(`AlQuran API error: ${res.status}`)

  const data = await res.json()
  const ayahs = data.data.ayahs

  const result: QuranPage = {
    number: page,
    ayahs: ayahs.map((a: { number: number; text: string; surah: { number: number; name: string; englishName: string } }) => ({
      number: a.number,
      text: a.text,
      surah: { number: a.surah.number, name: a.surah.name, englishName: a.surah.englishName },
    })),
    surahName: ayahs[0]?.surah?.name ?? '',
  }

  pageCache.set(page, result)
  return result
}

export function getPageText(page: QuranPage): string {
  return page.ayahs.map((a) => a.text).join(' ')
}

export function getPageWords(page: QuranPage): string[] {
  return getPageText(page)
    .split(/\s+/)
    .filter((w) => w.length > 0)
}

/**
 * Normalize Arabic text for comparison.
 * Removes diacritics and standardizes letters.
 */
export function normalizeArabic(text: string): string {
  return text
    // Remove diacritics (harakat)
    .replace(/[\u064B-\u065F\u0670]/g, '')
    // Normalize alef variants → ا
    .replace(/[أإآٱ]/g, 'ا')
    // Normalize hamza on waw
    .replace(/ؤ/g, 'و')
    // Normalize hamza on ya (without dots)
    .replace(/ئ/g, 'ي')
    // Normalize ta marbuta
    .replace(/ة/g, 'ه')
    // Remove tatweel
    .replace(/ـ/g, '')
    .trim()
}

File: backend/app/services/quran.py
import requests
import re
import logging

logger = logging.getLogger(__name__)

def get_quran_page_text(page_number):
    try:
        url = f"http://api.alquran.cloud/v1/page/{page_number}/quran-uthmani"
        response = requests.get(url, timeout=10)
        if response.status_code == 200:
            data = response.json()
            ayahs = data['data']['ayahs']
            return " ".join([ayah['text'] for ayah in ayahs])
        return None
    except Exception as e:
        logger.error(f"Error fetching Quran text: {e}")
        return None

def normalize_arabic(text):
    # Supprimer les diacritiques (Tashkeel)
    text = re.sub(r'[\u064B-\u065F\u0670]', '', text)
    # Normaliser les Alifs (أ, إ, آ -> ا)
    text = re.sub(r'[أإآ]', 'ا', text)
    # Normaliser les Ya (ى -> ي)
    text = re.sub(r'ى', 'ي', text)
    # Normaliser les Waw (ؤ -> و)
    text = re.sub(r'ؤ', 'و', text)
    # Supprimer le Tatweel (ـ)
    text = re.sub(r'ـ', '', text)
    # Supprimer la ponctuation, les chiffres et caractères non-alphanumériques (sauf espaces)
    text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'\d+', '', text)
    return text.strip()
